src/Sentry/MetricAggregator.cs

```diff
@@ -0,0 +1,273 @@
+            _ => throw new ArgumentOutOfRangeException(nameof(type), type, "Unknown MetricType")
+        };
+
+        timeBucket.AddOrUpdate(
+            GetMetricBucketKey(type, key, unit.Value, tags),
+            addValuesFactory,
+            (_, metric) =>
+            {
+                metric.Add(value);
+                return metric;
+            }
+        );
+    }
+
+    private async Task RunLoopAsync()
+    {
+        _options.LogDebug("MetricsAggregator Started.");
+
+        using var shutdownTimeout = new CancellationTokenSource();
+        var shutdownRequested = false;
+
+        try
+        {
+            while (!shutdownTimeout.IsCancellationRequested)
+            {
+                // If the cancellation was signaled, run until the end of the queue or shutdownTimeout
+                if (!shutdownRequested)
+                {
+                    try
+                    {
+                        var delay = Task.Delay(_flushInterval, shutdownTimeout.Token).ConfigureAwait(false);
+                        await Task.Delay(_flushInterval, shutdownTimeout.Token).ConfigureAwait(false);
+                    }
+                    // Cancellation requested and no timeout allowed, so exit even if there are more items
+                    catch (OperationCanceledException) when (_options.ShutdownTimeout == TimeSpan.Zero)
+                    {
+                        _options.LogDebug("Exiting immediately due to 0 shutdown timeout.");
+
+                        shutdownTimeout.Cancel();
+
+                        return;
+                    }
+                    // Cancellation requested, scheduled shutdown
+                    catch (OperationCanceledException)
+                    {
+                        _options.LogDebug(
+                            "Shutdown scheduled. Stopping by: {0}.",
+                            _options.ShutdownTimeout);
+
+                        shutdownTimeout.CancelAfterSafe(_options.ShutdownTimeout);
+
+                        shutdownRequested = true;
+                    }
+                }
+
+                // Work with the envelope while it's in the queue
+                foreach (var key in GetFlushableBuckets())
+                {
+                    // TODO: Check if a shutdown request has been made
+                    try
+                    {
+                        _options.LogDebug("Flushing metrics for bucket {0}", key);
+                        var bucket = Buckets[key];
+                        _captureMetrics(bucket.Values);
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        _options.LogInfo("Shutdown token triggered. Time to exit.");
+                        return;
+                    }
+                    catch (Exception exception)
+                    {
+                        _options.LogError(exception, "Error while processing metric aggregates.");
```

Comment:
When can we get here? I wonder if this could result in something throwing on each iteration and the code not stopping

Reply:
The `_captureMetrics` Action (which is the only thing that could trhow) is a dependency that gets passed into the `MetricAggregator`. Typically that would be this:
https://github.com/getsentry/sentry-dotnet/blob/c83e9c91840d0f78b070df426d318b43493f0cca/src/Sentry/SentryClient.cs#L244-L248

The most likely thing to go wrong is probably the [`Metric.SerializeAsync` call](https://github.com/getsentry/sentry-dotnet/blob/c83e9c91840d0f78b070df426d318b43493f0cca/src/Sentry/Protocol/Metrics/Metric.cs#L56) that happens when creating a `statsd` envelope item. 

Making the change on what you thought was a race condition above will prevent an infinite loop... and I think having the try catch here makes sense to make sure we don't crash the user's application. So I think once we always remove a bucket from the `Buckets` collection before trying to send the metrics through, we should be good.